homeassistant/components/media_player/dlna_dmr.py

```diff
@@ -0,0 +1,438 @@
+                               discovery_info=None):
+    """Set up DLNA DMR platform."""
+    # ensure this is a DLNA DMR device, if found via discovery
+    if discovery_info and \
+       'upnp_device_type' in discovery_info and \
+       discovery_info['upnp_device_type'] not in UPNP_DEVICE_MEDIA_RENDERER:
+        _LOGGER.debug('Device is not a MediaRenderer: %s, device_type: %s',
+                      discovery_info.get('ssdp_description'),
+                      discovery_info['upnp_device_type'])
+        return
+
+    if config.get(CONF_URL) is not None:
+        url = config[CONF_URL]
+        name = config.get(CONF_NAME)
+    elif discovery_info is not None:
+        url = discovery_info['ssdp_description']
+        name = discovery_info['name']
+
+    if DLNA_DMR_DATA not in hass.data:
+        hass.data[DLNA_DMR_DATA] = {}
+
+    # build upnp/aiohttp requester
+    from async_upnp_client.aiohttp import AiohttpSessionRequester
+    session = async_get_clientsession(hass)
+    requester = AiohttpSessionRequester(session, True)
+
+    # ensure event handler has been started
+    with (await ASYNC_LOCK):
+        # discovered components don't get default values in config
+        config[CONF_LISTEN_PORT] = config.get(CONF_LISTEN_PORT,
+                                              DEFAULT_LISTEN_PORT)
+        server_host, server_port = determine_listen_ip(config)
+        event_handler = await async_start_event_handler(hass,
+                                                        server_host,
+                                                        server_port,
+                                                        requester)
+
+    # create upnp device
+    from async_upnp_client import UpnpFactory
+    factory = UpnpFactory(requester, disable_state_variable_validation=True)
+    try:
+        upnp_device = await factory.async_create_device(url)
+    except (asyncio.TimeoutError, aiohttp.ClientError):
+        _LOGGER.debug('PlatformNotReady')
+        raise PlatformNotReady()
+
+    # wrap with DmrDevice
+    from async_upnp_client.dlna import DmrDevice
+    dlna_device = DmrDevice(upnp_device, event_handler)
+
+    # create our own device
+    device = DlnaDmrDevice(dlna_device, name)
+    _LOGGER.debug("Adding device: %s", device)
+    async_add_devices([device], True)
+
+
+class DlnaDmrDevice(MediaPlayerDevice):
+    """Representation of a DLNA DMR device."""
+
+    def __init__(self, dmr_device, name=None):
+        """Initializer."""
+        self._device = dmr_device
+        self._name = name
+
+        self._available = False
+        self._subscription_renew_time = None
+
+    async def async_added_to_hass(self):
+        """Callback when added."""
+        self._device.on_event = self._on_event
+
+        # register unsubscribe on stop
+        bus = self.hass.bus
+        bus.async_listen_once(EVENT_HOMEASSISTANT_STOP,
+                              self._async_on_hass_stop)
+
+    @property
+    def udn(self):
```

Comment:
This seems not needed. We can use `self._device.udn` directly in `unique_id` property.